### ISSUE TITLE
FR: Provide an API for developers which allows them to dial within overlay networks

### ISSUE THREAD

**Issue by eyedeekay** [tailscale/tailscale#8347]
### What are you trying to do?

I would like to set up a `headscale` server within the I2P network, then use that `headscale` server to orchestrate `tailscale` clients within the I2P network. I2P provides API's which third-party applications(which Tailscale would be for us) can use to connect to I2P without the need for any configuration on the user's part, they may simply supply an I2P address in lieu of an IP address when dialing a connection, or an I2P private key when listening on a connection. I would like to enable Tailscale to use these API's using the `onramp` library for overlay network connections.

### How should we solve this?

`onramp` provides standard implementations of Go network interfaces such as `net.Conn`, `net.PacketConn` and `net.Listener` and compatible functions for dialing connections and accepting on listeners. In order to enable library users of Tailscale(in this case, I2P) to communicate with I2P, it should be possible to "override" the underlying implementations of functions which return the network interfaces,`net.Conn` listener, e.g. `net.Dial`, `net.DialContext`, `net.Listen`.

### What is the impact of not solving this?

Without this change, it will remain impossible to use Tailscale over I2P without extraordinary workarounds. SOCKSification of tailscale does not appear to be a viable solution.

### Anything else?

 - https://github.com/eyedeekay/onramp
 - https://github.com/eyedeekay/epoccipital

**Comment by DentonGentry** [comment]
You're asking that Tailscale use onion routing networks to connect to other nodes on the tailnet.

**Comment by eyedeekay** [comment]
@DentonGentry not exactly, what I'm after is actually quite a bit more basic than that.

I don't think anything needs to change in Tailscale itself except to accommodate alternate implementations of some `net.*` interfaces when being used as a library. What I want is to be able to `import` Tailscale in another application, then "swap out" the dialer and listener functions used by Tailscale with ones provided by the library that interfaces with the onion-routed network. This would allow developers who want to(like myself) to create a version of a Tailscale-like which works over onion-routed networks only without requiring that Tailscale itself change how it works.
